src/product.cpp

```diff
@@ -186,6 +186,16 @@ size_t tsym::Product::hash() const
     return std::hash<BasePtrList>{}(ops);
 }
 
+unsigned tsym::Product::complexity() const
+{
```

Comment:
this must be implemented in BasePtrList, these lines identically occur multiple times throught types with an arbitrary number of operands, and this duplication should be avoided.